conda_project/project.py

```diff
@@ -3,12 +3,337 @@
+        else:
+            return False
+
+    @property
+    def is_prepared(self) -> bool:
+        """
+        bool: Returns True if the Conda environment exists and is consistent with
+              the environment source and lock files, False otherwise. If is_locked is
+              False is_prepared is False.
+        """
+        if (self.prefix / "conda-meta" / "history").exists():
+            if self.is_locked:
+                installed_pkgs = call_conda(
+                    ["list", "-p", str(self.prefix), "--explicit"]
+                ).stdout.splitlines()[3:]
+
+                lock = parse_conda_lock_file(self.lockfile)
+                rendered = render_lockfile_for_platform(
+                    lockfile=lock,
+                    platform=current_platform(),
+                    kind="explicit",
+                    include_dev_dependencies=False,
+                    extras=None,
+                )
+                locked_pkgs = [p.split("#")[0] for p in rendered[3:]]
+
+                return installed_pkgs == locked_pkgs
+
+        return False
 
-ENVIRONMENT_YAML_FILENAMES = ("environment.yml", "environment.yaml")
+    def lock(
+        self,
+        force: bool = False,
+        verbose: bool = False,
+    ) -> None:
+        """Generate locked package lists for the supplied or default platforms
+
+        Utilizes conda-lock to build the .conda-lock.yml file.
+
+        Args:
+            force:       Rebuild the .conda-lock.yml file even if no changes were made
+                         to the dependencies.
+            verbose:     A verbose flag passed into the `conda lock` command.
+
+        """
+        if self.is_locked and not force:
+            if verbose:
+                print(
+                    f"The lockfile {self.lockfile.name} already exists and is up-to-date.\n"
+                    f"Run 'conda project lock --force {self.name} to recreate it from source specification."
+                )
+            return
+
+        # Setup temporary file for conda-lock to write to.
+        # If a package is removed from the environment source
+        # after the lockfile has been created conda-lock updates
+        # the hash in the lockfile but does not remove the unspecified
+        # package (and necessary orphaned dependencies) from the lockfile.
+        # To avoid this scenario lockfiles are written to a temporary location
+        # and copied back to the self.lockfile path if successful.
+        tempdir = Path(tempfile.mkdtemp())
+        lockfile = tempdir / self.lockfile.name
+
+        channel_overrides, platform_overrides = self._overrides
+
+        specified_channels = []
+        for fn in self.sources:
+            env = EnvironmentYaml.parse_yaml(fn)
+            for channel in env.channels or []:
+                if channel not in specified_channels:
+                    specified_channels.append(channel)
+
+        with redirect_stderr(StringIO()) as _:
```

Comment:
On second thought, we may want to unify these various contexts into one, which would make the logic a lot cleaner.

```
with self.tmp_lockfile(self.lockfile) as fp:
   make_lock_files()...
```

This is what `atomicwrites` does I believe, or something similar, but we shouldn't take on a dependency for it.

If this sounds good, I'm happy to contribute a patch in another PR.

Reply:
Yes, let's do that in a separate PR: see #31